dashboard/test/controllers/api/v1/pd/workshop_enrollments_controller_test.rb

```diff
@@ -254,22 +254,38 @@ class Api::V1::Pd::WorkshopEnrollmentsControllerTest < ::ActionController::TestC
     refute_nil enrollment.code
   end
 
-  test 'CSF Intro enrollments can be created' do
-    assert_creates(Pd::Enrollment) do
-      post :create, params: {
-        workshop_id: @workshop.id,
-        school_info: school_info_params,
-        csf_intro_intent: 'Yes',
-        csf_intro_other_factors: [
-          "I want to learn computer science concepts.",
-          "I have available time on my schedule for teaching computer science."
-        ]
-      }.merge(enrollment_test_params)
-      assert_response :success
-      assert_equal RESPONSE_MESSAGES[:SUCCESS], JSON.parse(@response.body)["workshop_enrollment_status"]
-    end
-    enrollment = Pd::Enrollment.last
-    refute_nil enrollment.code
+  test 'can enroll in workshop when all params are submitted' do
```

Comment:
apologies if I am missing it, but can you also include a test which verifies that you cannot enroll when a required param is missing?

Reply:
Ah, I meant to include this in the description! We currently have no validation for dynamically-required fields –– I was wondering why all my tests were passing, and then I looked at the code. Unless it's a required field for all forms, like [email](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/models/pd/enrollment.rb#L61) or [school](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/models/pd/enrollment.rb#L67).

I'll add a ticket to track this work. I think we'd need to do something like we do in teacher application which adds required fields based on the program they've selected.

Reply:
interesting! right now it looks like we allow any parameter through as long as it is a valid param for any enrollment type, via `enrollment_params`. it also looks like the enrollment "type" is determined by looking at `enrollment.workshop.subject`. 

if the above is accurate, can you tell me if the following idea makes sense?
* define a map or helper method which maps enrollment.workshop.subject to a list of required params
* pass the selected list of required params to a new call to `params.require` (see [docs](https://guides.rubyonrails.org/action_controller_overview.html#strong-parameters)) in `enrollment_params`

I don't mean to prescribe a solution, but am partially asking to see if my understanding is correct.

Reply:
> I'll add a ticket to track this work. I think we'd need to do something like we do in teacher application which adds required fields based on the program they've selected.

is this what you are referring to?  https://github.com/code-dot-org/code-dot-org/blob/8e03fa188a6b2872b3643a1ea08e5d7cfadfb744/dashboard/app/models/pd/application/teacher_application.rb#L709

Reply:
Your idea makes a lot of sense!

In Teacher Application, we define "[dynamic_required_fields](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/models/pd/application/teacher_application.rb#L675)" which then gets used in a callback from Form in "[validate_required_fields](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/models/concerns/pd/form.rb#L80)".

I'm not exactly sure why we aren't using `params.require` in our application (although I do see it in [applications_controller](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/controllers/api/v1/pd/applications_controller.rb#L198), but your idea sounds much simpler.

Ticket for this work is at https://codedotorg.atlassian.net/browse/ACQ-417 and also linked above